pyiron_contrib/atomistics/runner/storageclasses.py

```diff
@@ -0,0 +1,353 @@
+# Copyright (c) Georg-August-Universität Göttingen - Behler Group
+# Distributed under the terms of "New BSD License", see the LICENSE file.
+"""Class Overrides for all `runnerase` storage classes.
+
+The ASE calculator for RuNNer, provided by the Python package `runnerase`,
+stores data like the values of symmetry functions or the weights of atomic
+neural networks in custom classes. However, these classes generally do not know
+how to write themselves to HDF5 storage format. Therefore, these classes are
+extended in this module with additional functions, typically `to_hdf(...)` and
+`from_hdf(...)`.
+
+Attributes
+----------
+     SymmetryFunctionSet : runnerase.SymmetryFunctionSet
+        Job class for generating and evaluating potential energy surfaces using
+        RuNNer.
+
+Reference
+---------
+    [RuNNer online documentation](https://theochem.gitlab.io/runner)
+"""
+
+from typing import Optional
+from abc import abstractmethod
+
+import numpy as np
+
+from runnerase.symmetryfunctions import SymmetryFunctionSet
+from runnerase.storageclasses import (RunnerSymmetryFunctionValues,
+                                      RunnerStructureSymmetryFunctionValues,
+                                      RunnerSplitTrainTest,
+                                      RunnerFitResults,
+                                      RunnerWeights, RunnerScaling)
+from pyiron_base import ProjectHDFio
+from pyiron_base.interfaces.has_hdf import HasHDF
+from pyiron_base import FlattenedStorage
+
+from .utils import pad, unpad
+
+
+class HDFSymmetryFunctionValues(FlattenedStorage):
+    """Extend runnerase RunnerSymmetryFunctionValues with HDF5 compatibility."""
+
+    __hdf_version__ = '0.3.0'
+
+    def from_runnerase(
+        self,
+        runnerase_sfvalues: RunnerSymmetryFunctionValues
+    ) -> None:
+        """Fill `self` with information of the corresponding `runnerase` object.
+
+        `runnerase` stores the symmetry function values of each structure in
+        a separate `RunnerStructureSymmetryFunctionValues` object. However,
+        it is very inefficient to read and write all of this information
+        separately into HDF5 storage. Therefore, we use a `FlattenedStorage`
+        where all the symmetry function values of all structures are stored
+        in one large flat array.
+
+        Matters are complicated further because `FlattenedStorage` only accepts
+        data chunks (read here: structures) where all different arrays of
+        information have the same length. Unfortunately, one can have different
+        numbers of symmetry functions for each element and varying numbers of
+        elements for each structure in a training dataset.
+
+        The solution here is to `pad` the sfvalues of each group of elements in
+        one structure with rows filled with `np.NaN`. This way, all arrays for
+        one structure have length `number_of_atoms` and can be efficiently
+        stored and retrieved even for large  datasets.
```

Comment:
Not sure I follow.  Let's say I have a structure with (Al, Al, Li) atoms, then a NNP might have 4 descriptors for Al, but only 2 for Li and therefore you want to store something like
```
Element |           SF_Al |   SF_Li
-----------------------------------
Al      |   1   2   3   4 | nan nan
Al      |   2   1   4   3 | nan nan
Li      | nan nan nan nan |   1   2
```
Correct?

Reply:
Yeah exactly. From `RunnerStructureSymmetryFunctionValues` (I know, I hate the name, too) I get one array of symmetry functions for each element. However, the length of each array is not the number of atoms in the structure but the number of atoms _of that element_ in the structure. In other words, it does not contain the `NaN` values from your table yet.

In order to store the symmetry functions at the right atom in `FlattenedStorage` I therefore have to add rows to each array. That is why I am calling the `pad` and `unpad` routines from `utils.py`.

Is that okay?

Reply:
Hm, for that kind of flexibility I can only think of long format tables right now, which is not what we want in `FlattenedStorage`.  So fine with me to leave this as is, until we have an idea.